presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -689,6 +716,27 @@ public void addInput(GroupByIdBlock groupIdsBlock, Page page)
             checkState(rawInputs != null && blockBuilders == null);
             rawInputs.ensureCapacity(rawInputsLength);
 
+            // Perform dedup for distinct aggregates
+            if (delegate instanceof DistinctingGroupedAccumulator) {
+                // Prepend the groupId block to the input page
+                Page withGroup = page.prependColumn(groupIdsBlock);
+
+                // Compute a mask for the distinct rows (including the group id)
+                checkState(markDistinctHash.isPresent());
+                Work<Block> work = markDistinctHash.get().markDistinctRows(withGroup);
+                checkState(work.process());
+                Block distinctMask = work.getResult();
+
+                // Filter out duplicate rows
+                Page dedupPage = filter(withGroup, distinctMask);
+
+                // Update GroupByIdBlock from distinctPage
+                groupIdsBlock = new GroupByIdBlock(groupIdsBlock.getGroupCount(), dedupPage.getBlock(0));
+
+                // Remove the groupId block from distinct Page
+                page = dedupPage.extractChannels(IntStream.range(1, dedupPage.getChannelCount()).toArray());
```

Comment:
Can this be conditional on the amount of deduplication?

Not that serious about trying to push to optimize this in the case where deduping isn't working, but worth raising the question.

Reply:
Oh this doesn't copy!